compaction/compaction.cc

```diff
@@ -1716,9 +1818,16 @@ static future<compaction_result> scrub_sstables_validate_mode(sstables::compacti
     clogger.info("Scrubbing in validate mode {}", sstables_list_msg);
 
     auto permit = table_s.make_compaction_reader_permit();
-    auto reader = sstables->make_crawling_reader(schema, permit, descriptor.io_priority, nullptr);
 
-    const auto validation_errors = co_await scrub_validate_mode_validate_reader(std::move(reader), cdata);
+    uint64_t validation_errors{0};
+
+    {
+        auto reader = sstables->make_crawling_reader(schema, permit, descriptor.io_priority, nullptr);
+        validation_errors += co_await scrub_validate_mode_validate_reader(std::move(reader), cdata);
+    }
+    for (auto sst : descriptor.sstables) {
+        validation_errors += co_await scrub_validate_mode_validate_index(sst, permit, descriptor.io_priority, cdata);
+    }
```

Comment:
Why don't we merge both validation methods into a single one?

Looks like scrub_validate_mode_validate_index() has the potential to apply the validator when iterating through fragments for a given partition retrieved from the index entry. Then we'd be stressing both index and data, using a single method, reducing the operation time.

Reply:
If the index has wholes in it, we'd be missing validation data for those partitions. We have to validate the index and data separately to avoid corruptions that stay hidden when looking at both index+data.

Reply:
Makes sense.

Reply:
Agree with @raphaelsc first comment. We don't need to apply the results of the index read to the data read, just co-scan the data and index files, and make sure that every index entry that we see has a corresponding entry in the data file and the values match. It also allows us to verify the promoted index.